tests/static-analysis/CollectionT/append.php

```diff
@@ -0,0 +1,97 @@
+//
+// Note that analysers can behave slightly differently, notably PHPStan does not consider
+// multiple parameters passed too append() as potentially part of a union type.
+
+//## SUCCESS ###
+appendT_checkList(CollectionT::empty()->append(1));
+appendT_checkList(CollectionT::empty()->append(1, 2));
+appendT_checkList(CollectionT::empty()->append(...[1, 2]));
+appendT_checkList(CollectionT::fromIterable([5])->append(2));
+appendT_checkList(CollectionT::fromIterable([5])->append(1, 2));
+appendT_checkList(CollectionT::fromIterable([5])->append(...[1, 2]));
+
+// Analysers sometime need to know the type of the first item in the collection or the appended one
+// otherwise they might narrow it down too much, i.e. only allow array{foo: int} or array{bar: 2}
+/** @var array<string, int> $foo */
+$foo = ['foo' => 1];
+/** @var array<string, int> $bar */
+$bar = ['bar' => 2];
+appendT_checkListWithMap(CollectionT::empty()->append($foo));
+appendT_checkListWithMap(CollectionT::empty()->append($foo, ['bar' => 2]));
+appendT_checkListWithMap(CollectionT::empty()->append(...[$foo, $bar]));
+appendT_checkListWithMap(CollectionT::fromIterable([1 => $foo])->append($bar));
+
+// These should work but they don't due to the way analysers interpret empty()
+// or variadic parameters of different types passed to append()
+/** @psalm-suppress InvalidArgument @phpstan-ignore-next-line */
+appendT_checkMixed(CollectionT::empty()->append(1, '2'));
+/** @psalm-suppress InvalidArgument */
+appendT_checkMixed(CollectionT::empty()->append(...[1, '2']));
+/** @phpstan-ignore-next-line */
+appendT_checkMixed(CollectionT::fromIterable(['foo' => 1, 'bar' => '2'])->append(1, '3'));
+
+//## VALID FAILURE ###
+appendT_checkList(CollectionT::empty()->append(1, 'foo')); // @phpstan-ignore-line
+appendT_checkList(CollectionT::fromIterable([5])->append('foo')); // @phpstan-ignore-line
+appendT_checkList(CollectionT::fromIterable([5])->append('foo', 1)); // @phpstan-ignore-line
+
+/** @psalm-suppress InvalidScalarArgument @phpstan-ignore-next-line */
+appendT_checkListWithMap(CollectionT::empty()->append($foo, ['bar' => 'baz']));
+/** @psalm-suppress InvalidScalarArgument @phpstan-ignore-next-line */
+appendT_checkListWithMap(CollectionT::fromIterable([1 => $foo])->append(['bar' => 'baz']));
+
+/**
+ * Append will transform any CollectionT<string, int> into CollectionT<int|TKey, int>.
+ *
+ * @psalm-suppress InvalidScalarArgument
+ * @phpstan-ignore-next-line
+ */
+appendT_checkMap(CollectionT::fromIterable($foo)->append(3));
+
+/** @psalm-suppress InvalidArgument @phpstan-ignore-next-line */
+appendT_checkMixed(CollectionT::empty()->append(1, [3]));
+/** @psalm-suppress InvalidArgument */
+appendT_checkMixed(CollectionT::empty()->append(...[1, [3]]));
+/** @psalm-suppress InvalidArgument @phpstan-ignore-next-line */
+appendT_checkMixed(CollectionT::fromIterable(['foo' => 1, 'bar' => '2'])->append(1, ['3']));
```

Comment:
@drupol have a look at this file comparing with the [other one](https://github.com/loophp/collection/pull/102/files?diff=split&w=1#diff-735f11d2c95768f8eed8f60feffd52f4225859f05c6a9c526695ef30108fbf23R1) (mixed collection) to understand the difference in what `CollectionT`'s interface would allow. Even though it doesn't yet have any runtime difference (doesn't enforce a single type), static analysis would stop the user from adding mixed types to it. Also note the difference in the `@var` annotations needed (or not needed for the typed version).

Reply:
This is beautiful, it should be the default!

Reply:
So basically your idea for making this happening is to not enforce any type? Just enforce the type through the annotations?
The consequence of this will be a change in operations, making them non-variadic. Right?

Reply:
I think your **"Idea 1"** from the discussion https://github.com/loophp/collection/discussions/99#discussion-3381462 is still valid and would be useful. It should be used _in conjunction_ with the type annotations, but it might be limited due to what `gettype` can do, unless we use it recursively and check all the keys and value types in the case where you have a collection of arrays.

Also I am not sure about **union types** because static analysis will allow instantiation, as mentioned [here](https://github.com/loophp/collection/pull/102#discussion_r642100624), but the runtime enforcement solely based on `gettype` will throw an exception. Perhaps it wouldn't be an issue though as long as the user knows about the behaviour.

Regarding the **operations** I don't think there would be any implications in how they work, they would still be variadic. See the examples in the static analysis checks for `append` - it still allows the user to append multiple elements; the only inconvenience might be that _sometimes_ extra manual type hints might be needed on the user side to satisfy the static analysis. But this will also be dependant on the strictness levels they use for psalm, phpstan

Reply:
Well, the idea is to keep this library simple, well documented. If we start shipping 2 collection classes, it will confuse people.

What I think is the following:

We should update `Collection.php` and use proper annotations, as if it was a typed collection, without enforcing it.

Then we could provide a new operation (**Idea 2** from #99) if we want to enforce it.

About the `gettype`, I don't think this is the valid way. Let's say you want to enforce a particular type of object (_implementing a specific interface_), when using `gettype` only, it's simply not possible because it will always return 'object'. 

WDYT?

Reply:
> We should update Collection.php and use proper annotations, as if it was a typed collection, without enforcing it.

I would like that but I thought we established it's not possible if we want to allow for the flexibility to mix types. And because of this 2x different interfaces are needed for the operations (at least for some of them, maybe not for all). Let's look at `Appendable`:

A) Typed:
```PHP
/**
 * @template TKey
 * @template T
 */
interface AppendableT
{
    /**
     * Add one or more items to a collection.
     *
     * @template U of T
     *
     * @param U ...$items
     *
     * @return CollectionT<int|TKey, T|U>
     */
    public function append(...$items): CollectionT;
```

B) Mixed:
```PHP
/**
 * @template TKey
 * @template T
 */
interface Appendable
{
    /**
     * Add one or more items to a collection.
     *
     * @param mixed ...$items
     *
     * @return Collection<int|TKey, T|mixed>
     */
    public function append(...$items): Collection;
}
```

Both are valid and have their own use cases. A) only allows you to append elements of the same type that you already have in the collection, whereas B) allows you to append anything you want.

If you like the idea of having a single collection class more then would you be happy to restrict the way it can be used?

Reply:
I understand for sure.

So to summarize, in the end, we will end up with 2 classes (`Collection` and `CollectionT` for example) with exactly the same PHP code, but with different annotation?

Reply:
actually looking at this PR and trying out a few things it looks like we won't need to duplicate the code for the operations, only the interfaces would have to be different.

So if this is indeed correct then to summarise we would have:

`Collection`
- implements `src/Contract/Collection`, which extends all the current operation interfaces but modified a bit to allow for the mixed usage without code analysis errors
- uses the operations in `src/Operation` just like it does now

`CollectionT`
- implements `src/Contract/CollectionT`, which extends new operation interfaces `AppendableT`, `ApplyableT` with updated type hints
- uses the same operations in `src/Operation`
- enforces basic types at runtime, i.e. primitives like `int`, `string`, `array`. For objects we can use `get_class` to ensure the same class. I think this is important for the end user because it justifies a bit more choosing which collection to use. When you _know_ you want the same type you don't want to solely rely on static analysis because that can be worked around/ignored

In essence the only code that would be actually duplicated would be the methods themselves in `src/Collection.php` and `src/CollectionT.php`. One thought I had was to put them on an abstract class but due to requiring different interfaces it's not possible

Reply:
@drupol hey, sorry I forgot to tag you here 😄. What do you think, are you on board with this?
I could add the runtime type enforcing iterator to this PR and it can be the start of CollectionT

Reply:
> One thought I had was to put them on an abstract class but due to requiring different interfaces it's not possible.

Do you think we could have the same interface for both type of Collection? Wouldn't it be better?

Reply:
I'm a bit confused 😓. Are you saying we should update the Collection interface to not allow mixing types? Because if we are to maintain the current flexibility it makes sense to have two separate interfaces like I described [here](https://github.com/loophp/collection/pull/102#discussion_r642635206).

Reply:
Ok... I don't like the idea of duplicating each interfaces. I'm thinking about something simpler.
I guess that as long as we will have 2 Collections, we must have twice all the interfaces. 
Do you see a better solution?

Reply:
I see. I think the separate interfaces would property show the _intent_ behind each collection. 

The only simpler solution that I can think of if we want to avoid this is: treat `Collection` like a `CollectionT` in the type system, but not at runtime. It would have to be specified in the documentation and when you'd want to use it like a mixed collection you might sometimes need to use annotations to satisfy the static analysis. I think it's not a bad idea, in this case we can ditch CollectionT altogether and just have the `strict` operation you mentioned in Idea 2

Reply:
Yes, this is is exactly what I have in mind. If you're ok with that, we can go on. Deal ? :)

Reply:
alright, let's give it a shot 😁. I'll refactor this PR to get rid of CollectionT and start some work towards the plan

Reply:
Excellent. Let me know if I can be of any help on this.

Reply:
I do have a question for you actually. What would you say about this implementation of `strict` (note: haven't tested yet, just something that suddenly came to mind):

```PHP
public function strict(): CollectionInterface 
{
    return self::fromCallable(static fn() => yield from (new TypedIterator($this->getIterator()));
}
```

This way we don't need a new operation and the logic is in the `TypedIterator` like in _Idea 1_. And the user doesn't need to provide any callback to `strict` because the logic is internal.

Reply:
This seems to be at first sight a possible solution. I'll test tomorrow, I'm about to sleep here.